### ISSUE TITLE
ordered list digits cut off

### ISSUE THREAD

**Issue by jgreely** [laurent22/joplin#4715]
The digits in ordered lists extend into the margin and get cut off. The problem is much more severe on the iOS client, especially with wider fonts, so that "monospace" won't even show a single-digit number. On the Mac desktop client, it will at least display correctly until you hit 100 items in the list.

## Environment

Joplin version: 10.7.2
Platform: iOS (iPhone XS & iPad Mini)
OS specifics: 14.4.1

## Steps to reproduce

1. Create note or todo containing ordered list with 10 or more rows.
2. On iOS, for items 10 and up, the first digit is cut off. 
3. Add a DIV tag at the top with `style="padding-left:16px"` to make the tens digit visible.
4. Add additional items to the list to take it above 100 rows, and the leading digit is again cut off.
5. Increase the padding to `32px` to make the hundreds digit visible.

## Describe what you expected to happen

All digits should be visible regardless of font or length of list.

![ol-default-font](https://user-images.githubusercontent.com/5074445/111885041-9300f780-8982-11eb-98a7-0649e3782e08.png)
![ol-monospace-font](https://user-images.githubusercontent.com/5074445/111885042-972d1500-8982-11eb-8809-30a4c8ded3bd.png)

**Comment by PIG208** [comment]
It appears that the same problem occurs to the desktop application on Windows as well when the size of the numbered list goes to 1000 or above.
![image](https://user-images.githubusercontent.com/39874143/111905751-32f47900-8a88-11eb-9ae2-3c6441cd8f50.png)
![image](https://user-images.githubusercontent.com/39874143/111905756-3be54a80-8a88-11eb-9dca-0a7b6e04ea62.png)

**Comment by CalebJohn** [comment]
I just tested this on v1.7.11 Ubuntun 20.04 and v1.7.5 Android 11.

I could duplicate this behaviour in the viewer on android and on the WYSIWYG editor on the desktop. I could not duplicate this behaviour on the markdown editor/viewer on desktop.

**Comment by tessus** [comment]
I was able to reproduce the problem on 

```
Joplin 1.8.0 (dev, darwin)

Client ID: e354a932e6e143c782264aad4112b674
Sync Version: 2
Profile Version: 34
Keychain Supported: Yes

Revision: 5b65186b4 (dev)
```

Although not at 100, but at 1,000.

<img width="175" alt="image" src="https://user-images.githubusercontent.com/223439/112092992-1a12c300-8b6f-11eb-92d4-1fecad85f1e6.png">

**Comment by jalajcodes** [comment]
I can also reproduce it on Windows after 1000 items.
```
Joplin 1.7.11 (prod, win32)

Client ID: a0c954327e82488cb38da36c5c2a167d
Sync Version: 2
Profile Version: 34
Keychain Supported: Yes

Revision: f560563d8 (master)
```
I am investigating this issue.

**Comment by singhakshita** [comment]
In noteStyle.ts , `margin-left : 1.7em ' is fixed , changing that  to 2em fixes the issue ,but  i guess making this dynamic  will be a better choice .
can i work on this issue

**Comment by CalebJohn** [comment]
@singhakshita Sure, go ahead. Be sure to read [the guidelines](https://joplinapp.org/gsoc2021/pull_request_guidelines/) first.

**Comment by coderrsid** [comment]
Is this solved? May i work on this?

**Comment by singhakshita** [comment]
no i haven't send a pull request yet , i m still working on it .

**Comment by coderrsid** [comment]
Well @singhakshita, i can reproduce this currently too on macOS.
Although what i figured out was that it takes all the elements in a single list item after 99th element, that's why it doesn't add space.

**Comment by singhakshita** [comment]
yeah i said that only , i was just clarifying that it might look like its adding space but actually its adding list items inside that 99th element

**Comment by coderrsid** [comment]
Although @tessus , if i try to set a left margin to the ordered list to fix this, should i consider list upto 1000 or even greater? Thanks.

**Comment by tessus** [comment]
I'm not really sure, but IMO this should be dynamic. I'm not a UI designer, but from a coding standpoint a dynamic solution would be preferable. In my case it only happens afer the 99th item and I never have numbered lists that long.

**Comment by jgreely** [comment]
On a desktop Chrome-based browser on a Mac, I can reproduce the 1000-length list problem with straight HTML, so I think correctly rendering three-digit list numbers on all platforms is sufficient. Experimentally, I've found that adding `ol li {padding-left:1.5rem}` does the trick on Mac/iPhone/iPad, even with large monospace fonts.

**Comment by PIG208** [comment]
** ghostwriter uses `list-style-position: inside;` **
![image](https://user-images.githubusercontent.com/39874143/122196108-83357880-cec9-11eb-8d6d-5cdd35709cec.png)
```css
ol, ul {
    list-style-position: inside;
    padding: 0;
}
```
This can be the correct solution for us if we would like to change the way list items align if we want to support as many digits as we want.

** Github uses a fixed padding of 2rem: **
![image](https://user-images.githubusercontent.com/39874143/122196429-d1e31280-cec9-11eb-96a1-dad8a6b73998.png)
```css
.markdown-body ol, .markdown-body ul {
    padding-left: 2em;
}
```
Apparently, this runs into the same problem when it comes to 1000, but generally, this is fine because they keep the markdown in a container with a padding of 32px.
![image](https://user-images.githubusercontent.com/39874143/122196721-0eaf0980-ceca-11eb-940c-7e46228302e8.png)

I think the Github solution would be totally feasible in our case. By simply selecting an appropriate margin/padding we can support up to 99999 list items, which should cover most of the use case (plus it might be noticeably unresponsive when a list reaches that many items, which is a more critical problem than the cut off). A dynamic solution might not be necessary.

**Comment by harsh0032** [comment]
May i work on this ?

**Comment by CalebJohn** [comment]
@harsh0032 go ahead.

**Comment by Ritesh-Aggarwal** [comment]
Is this issue solved?

**Comment by roman-r-m** [comment]
> Is this issue solved?

Doesn't seem so. You can try the latest pre-release and see for yourself.

**Comment by Ritesh-Aggarwal** [comment]
I checked and the issue still persists. Can i work on this?

**Comment by PIG208** [comment]
@Ritesh-Aggarwal you can go ahead!

**Comment by harsh0032** [comment]
Hi i am not working on this anymore

**Comment by Ritesh-Aggarwal** [comment]
I am trying to setup for development but i am getting errors when i run npm install in root on windows
I checked the issue and found this forum but it isn't still working.Can anyone refer to a better solution to this issue.
forum mentioned: https://discourse.joplinapp.org/t/error-while-setting-up/15157/6

**Comment by JackGruber** [comment]
@Ritesh-Aggarwal please create a post on discourse for your installation problem of the dev env.

**Comment by Xindi7** [comment]
Hi. I just confirmed the issue yesterday at my Android and IOS devices. I could not go over 100. Is there any work being done on this issue?

**Comment by MuhammadBilalBinAtique** [comment]
Hi, I have examined the issue in both android and window 10.
In android, the problem occurs for 3 digit numbers ( after 99 ):
![joplin android bug](https://user-images.githubusercontent.com/74999164/144736368-30c1562a-4d01-4aeb-a532-d3b3f8f683c1.jpeg)
In window 10, the problem occurs for 4 digit numbers ( after 999 ):
![jolpin desktop bug](https://user-images.githubusercontent.com/74999164/144736387-d3d659cd-ec6f-4473-9658-461d11d9e695.PNG)
What I want to ask is **Can I work on it**?

**Comment by CalebJohn** [comment]
@MuhammadBilalBinAtique go ahead, but please check the previous attempts to fix this and make sure your solution is better. This is a complex problem and only correct solutions will be accepted.

**Comment by MuhammadBilalBinAtique** [comment]
@CalebJohn Can you please share some views on what the correct solution will look like?

**Comment by CalebJohn** [comment]
> @CalebJohn Can you please share some views on what the correct solution will look like?

Sorry @MuhammadBilalBinAtique, I have not looked into this problem deeply and cannot guide you in solving it.

**Comment by MuhammadBilalBinAtique** [comment]
@CalebJohn  While trying to run the desktop app, I am getting the following error:
![joplin error](https://user-images.githubusercontent.com/74999164/144780790-1ef5d1dd-1124-4edf-9331-5d0299f26093.PNG)

**Comment by tessus** [comment]
@MuhammadBilalBinAtique please do not post text as screenshots. seriously how do people come up with idiotic idea?

And then please maybe search for the problem on the forum or ask there. gh is used to discuss reproducible errors. The problem is known, nothing left to discuss. So either you can come up with a PR or use the forum for questions.

**Comment by jooji-san** [comment]
observations by @PIG208 are very helpful. We can use this css property to avoid hard-coding margins:
```
ol {
    list-style-position: inside;
}
```
 In #4821 @laurent22 was interested in what other markdown editors did. I checked GitHub and VSCode and turns out they also can't handle big numbers elegantly. They largely avoid this issue as @PIG208 already stated by having a padding.

**Comment by jooji-san** [comment]
The snippet below renders in browser (Firefox) by default like this (the same applies for Joplin):
```
<div>
  here's some text
  <ol>
    <li>item 1</li>
    <li>item 2</li>
  </ol>
  <ul>
    <li>item 1</li>
    <li>item 2</li>
  </ul>
</div>
```
![image](https://user-images.githubusercontent.com/11631221/147879940-14270196-2f9a-49bc-ad26-2b3a07dae6ba.png)

If we have large numbers: 
![image](https://user-images.githubusercontent.com/11631221/147879949-d23fc81d-bbdf-448a-8b01-1a47a18e0916.png)

If we use above mentioned property:
```
ol {
    list-style-position: inside;
}
```
![image](https://user-images.githubusercontent.com/11631221/147879974-2c12f771-bd26-40c4-9674-af10eff981af.png)

However, as you can see there is a mismatch in indentation between ordered and unordered lists. To remedy this we can either apply the same `list-style-position` property to `ul` also and set both `ul` and `ol` padding to 0:
![image](https://user-images.githubusercontent.com/11631221/147880075-8ed1b2e3-c266-4625-b599-a52962bcc825.png)
Or we could just set the `ol` padding so that it matches the indentation of the `ul`. 25px worked nicely in the browser for instance (`ul` padding is also set in pixels by default).:
![image](https://user-images.githubusercontent.com/11631221/147880226-5e50567b-5d22-44c7-9a0f-478c8ed9f6fc.png)
I prefer the second way as we don't alter the styling of the unordered list as it is not subject of this issue.

**Comment by jooji-san** [comment]
I will start working on the PR when we will decide which way to go.

**Comment by niharikamahajan02** [comment]
Is this issue still open? Anyone working on it?

**Comment by jooji-san** [comment]
I am waiting for a response from the contributors.

**Comment by CalebJohn** [comment]
@jooji-san I think it doesn't hurt to change the styling for ul along with ol, as long as they have the same behaviour.

I don't know if you've deeply gone into the code yet, but there is [one place](https://github.com/laurent22/joplin/blob/46438a588881f4cbedb1aea53b90089097869fc5/packages/renderer/noteStyle.ts#L147) where the styling for both is applied together, I think it makes sense to apply this change there. You can change the `list-style-position` to `inside`, and then adjust the margin all from the same place.

That seems to be the correct fix, but it does suffer from a change in the visuals. Personally I think it will be okay, but I'm certain that some users won't be happy with it. Perhaps this is something that can be put to a vote on the forum?

## Before
<img width="108" alt="image" src="https://user-images.githubusercontent.com/2179547/148191510-5a5c606b-c8c8-4003-b4f0-63c5708d8bf6.png">

## After
<img width="108" alt="image" src="https://user-images.githubusercontent.com/2179547/148191444-970f6e94-41e5-4334-b558-0cef107e7570.png">
(notice how it's not the first character in the numbers that is aligned, rather than the last)

**Comment by jooji-san** [comment]
yes, the fix is very easy to implement, but it may annoy people, so voting is fair.

**Comment by SayanDeveloper** [comment]
Sir, what if I will apply a check using javascript that how many elements are there? on the basis of that style can be modified.
Though this might create a little bit of time complexity.

**Comment by EshanSingh-ES** [comment]
Is this a UIKit app? If yes where can I find the Xcode project for this to reproduce this issue ? I am a beginner and have never contributed before

**Comment by VitoPareto** [comment]
Is someone currently working on this issue?

**Comment by namannangia** [comment]
Can i work on it ?

**Comment by roman-r-m** [comment]
> Can i work on it ? 

Sure

**Comment by Ankitvnsbihar** [comment]
hey i want to work on this problem can you assign me?

**Comment by CalebJohn** [comment]
@Ankitvnsbihar somebody else asked to work on it 2 days ago (see above). Please wait for them to finish or hand off the task.

**Comment by tessus** [comment]
We have explained many times that we do not assign issues. One of the reasons is that almost never a progress is made. A person asks if they can work on X and even if there's a reply that says "yes", you get a reply "thanks" but that was it. Weeks later.... No PR. Nothing. Nada. Zilch.

Now imagine this was assigned. It just doesn't make sense. Assignment only makes sense for core project members. And even then we rarely do that.

Thus, if you want to work on somehing, work on it. Create a valid PR that includes test cases for non GUI code and a manual testing process log for UI related code.

P.S.: Just look how man people asked if they could work on this. As you can see, the issue is still open, which means NOBODY actually did create a valid and correct PR.

**Comment by Ankitvnsbihar** [comment]
Hey bro i am new into open source can you help to(how can i get a code) to
solve any issue.
Please help.
Thank you.

**Comment by Daeraxa** [comment]
@laurent22 I've added a medium label just to stop the stale bot, feel free to change it if you need

**Comment by laurent22** [comment]
Good call, we still would like that to be fixed some day but it wasn't a good GFI

**Comment by dmnsea** [comment]
UPD: for me in version 3.0.8 for android ordered list is displayed properly now

~~Facing something similar, but in android version - numbers in range 0-99 works fine in both editor and view mode. numbers 100+ are okay while in editor mode, but in view mode it became cut off, so 100 turning into 00 and so on
![screenshot](https://i.imgur.com/hJVgPiM.png)~~
